design/baremetal-operator/limit-hosts-provisioning.md

```diff
@@ -0,0 +1,121 @@
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# limit-hosts-provisioning
+
+## Status
+
+provisional
+
+## Summary
+
+This proposal aims to introduce a limit to the number of hosts that the
+BareMetal Operator can provision or register simultaneously as a safety
+measure to avoid overloading excessively the system
+
+## Motivation
+
+In the current implementation BMO takes care of handling the complete
+provisioning cycle for a bare metal host without considering the pressure
+level applied to the underlying Ironic services (DHCP included).
+A user that desires to scale up several hundreds of nodes in a single step
+could suffer issues produced from the sudden resources depletion
+
+### Goals
+
+1. Avoid any system instability due the excessive number of hosts being
+   provisioned simultaneously
+
+### Non-Goals
+
+1. Optimize the deployment time of large batches of nodes
+2. Allow the user to configure such behavior
+3. Modify the current Ironic service layout
+
+## Proposal
+
+BMO reconciliation loop already checks for Ironic availability before
+processing an update and the request gets rescheduled if Ironic is not ready.
+After that, a new check will verify if Ironic is overloaded or not so that
+the request could be rescheduled or processed as usual.
+
+### Implementation Details/Notes/Constraints
+
+The new check will verify the number of currently hosts being provisioned/
+registered and, based on a configurable threshold (default set to 20), will
+determine if the current update can be processed or needs to be rescheduled.
+The difficult part consists in determining the number of hosts currently being
+provisioned, since Ironic does not offer any way for querying about its
+workload.
+The basic idea is to identify those nodes that already passed DHCP/PXE -
+considering it the most limiting factor for concurrent provisioning - by
+selecting the nodes that are in a waiting state and have
+_driver\_internal\_info_ field set by the agent.
+Following a simplified pseudo-code:
+
+```bash
+provisioning_slots = GetEnv("MAX_CONCURRENT_PROVISIONING_HOSTS", 20)
+
+for each Node
+  if Node.provision_state in
+    ("cleaning", "clean wait", "inspecting",
+    "inspect wait", "deploying", "wait call-back", "rescuing",
+    "rescue wait", "unrescuing", "deleting")
+    &&
+    Node.driver_internal_info.agent_url is present and not empty
+  then
+    provisioning_slots--
+
+if provisioning_slots < 0 then
+  reschedule current update
+
+```
+
+### Risks and Mitigations
+
+In a worst case scenario we could have a high number of “bad” hosts under
+provisioning but into an error state not recoverable. If such number is equal
+to the configured threshold, then the “bad” hosts could prevent other hosts
+to be correctly provisioned. A fair scheduling approach will be required
+to ensure that the available provisioning slots will be distributed evenly
+(the recently added _BMH ErrorCount_ field could be used for that)
```

Comment:
https://github.com/metal3-io/ironic-image/issues/203 made me think a little more about this proposal to use ironic's state to control the number of simultaneous operations we allow.

There's a hidden assumption in this design that the throttle limit is always going to be higher than the concurrency level of the baremetal-operator. If that is not true, then when we query Ironic and find slots available, we might start too many operations at once because we have too many concurrent threads thinking they have a slot available. We should document this assumption here as another risk. I don't think we should try to protect against it in the source code in any way.

Reply:
That's a good point and I agree it's better to document such assumption, I'll update the proposal. In the code, we could anyhow add a warning log trace when `BMO_CONCURRENCY` is greater than the throttle limit